rclpy/src/rclpy/clock.cpp

```diff
@@ -0,0 +1,244 @@
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+#include <rcl/rcl.h>
+#include <rcl/time.h>
+#include <rcl/types.h>
+
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+
+#include "time_point.hpp"
+
+using pybind11::literals::operator""_a;
+
+namespace rclpy
+{
+/// Destructor for a clock
+static void
+_rclpy_destroy_clock(void * p)
+{
+  auto clock = static_cast<rcl_clock_t *>(p);
+  if (!clock) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to get clock pointer in destructor");
+    rcl_reset_error();
+    return;
+  }
+
+  rcl_ret_t ret_clock = rcl_clock_fini(clock);
+  PyMem_Free(clock);
+  if (ret_clock != RCL_RET_OK) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini clock: %s",
+      rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+}
+
+py::capsule
+create_clock(rcl_clock_type_t clock_type)
+{
+  auto deleter = [](rcl_clock_t * ptr) {_rclpy_destroy_clock(ptr);};
+  auto clock = std::unique_ptr<rcl_clock_t, decltype(deleter)>(
+    static_cast<rcl_clock_t *>(PyMem_Malloc(sizeof(rcl_clock_t))),
+    deleter);
+  if (!clock) {
+    throw std::bad_alloc();
+  }
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcl_ret_t ret = rcl_clock_init(clock_type, clock.get(), &allocator);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to initialize clock");
+  }
+
+  PyObject * pyclock_c =
+    rclpy_create_handle_capsule(clock.get(), "rcl_clock_t", _rclpy_destroy_clock);
+  auto pyclock = py::reinterpret_steal<py::capsule>(pyclock_c);
+  // pyclock now owns the rcl_clock_t
+  clock.release();
+  return pyclock;
+}
+
+py::capsule
+clock_get_now(py::capsule pyclock)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  int64_t nanoseconds;
+  rcl_ret_t ret = rcl_clock_get_now(clock, &nanoseconds);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get current clock value");
+  }
+
+  return create_time_point(nanoseconds, clock->type);
+}
+
+bool
+clock_get_ros_time_override_is_enabled(py::capsule pyclock)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  bool is_enabled;
+  rcl_ret_t ret = rcl_is_enabled_ros_time_override(clock, &is_enabled);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get ROS time override status");
+  }
+
```

Comment:
Should this function be checking `PyErr_Occurred` as well?

Reply:
I don't think so. Checking is only necessary when arbitrary python code can be run, like time jump callbacks. [`rcl_is_enabled_ros_time_override()` doesn't invoke arbitrary python code](https://github.com/ros2/rcl/blob/4740c82864518a331ae98799f25b2ba085b22473/rcl/src/rcl/time.c#L333).